internal/provider/aws/health.go

```diff
@@ -0,0 +1,254 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/route53/route53iface"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	externaldns "sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/rs/xid"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/provider"
+)
+
+const (
+	idTag = "kuadrant.dev/healthcheck"
+
+	defaultHealthCheckPath             = "/"
+	defaultHealthCheckPort             = 80
+	defaultHealthCheckFailureThreshold = 3
+)
+
+var (
+	callerReference func(id string) *string
+)
+
+type Route53HealthCheckReconciler struct {
+	client route53iface.Route53API
+}
+
+var _ provider.HealthCheckReconciler = &Route53HealthCheckReconciler{}
+
+func NewRoute53HealthCheckReconciler(client route53iface.Route53API) *Route53HealthCheckReconciler {
+	return &Route53HealthCheckReconciler{
+		client: client,
+	}
+}
+
+func (r *Route53HealthCheckReconciler) Reconcile(ctx context.Context, spec provider.HealthCheckSpec, endpoint *externaldns.Endpoint, probeStatus *v1alpha1.HealthCheckStatusProbe) (provider.HealthCheckResult, error) {
+	healthCheck, exists, err := r.findHealthCheck(ctx, probeStatus)
+	if err != nil {
+		log.Log.Info("error finding health check", "err", err)
+		return provider.HealthCheckResult{}, err
+	}
+
+	address, _ := provider.GetEndpointAddress(endpoint)
+
+	if exists {
+		log.Log.Info("health check exists, updating")
+		status, err := r.updateHealthCheck(ctx, spec, endpoint, healthCheck)
+		if err != nil {
+			return provider.HealthCheckResult{}, err
+		}
+
+		return provider.NewHealthCheckResult(status, *healthCheck.Id, address, *spec.Host, ""), nil
+	}
+
+	log.Log.Info("health check doesn't exist, creating")
+	healthCheck, err = r.createHealthCheck(ctx, spec, endpoint)
+	if err != nil {
+		return provider.HealthCheckResult{}, err
+	}
+
+	return provider.NewHealthCheckResult(provider.HealthCheckCreated, *healthCheck.Id, address, *spec.Host, fmt.Sprintf("Created health check with ID %s", *healthCheck.Id)), nil
```

Comment:
what is a `HealthCheckResult` in this context? Is the result of the update/create or is it whether the health check is failing / passing?

Reply:
yeah, it's the result of the API call.